geom/twkb_parser.go

```diff
@@ -0,0 +1,660 @@
+
+// UnmarshalTWKB parses a Tiny Well Known Binary (TWKB), returning the
+// corresponding Geometry.
+func UnmarshalTWKB(twkb []byte, opts ...ConstructorOption) (Geometry, error) {
+	p := newtwkbParser(twkb, opts...)
+	g, err := p.nextGeometry()
+	return g, p.annotateError(err)
+}
+
+// UnmarshalTWKBWithHeaders parses a Tiny Well Known Binary (TWKB),
+// returning the corresponding Geometry, and any bounding box and any IDs
+// listed in its header information.
+//
+// If there is a bounding box header, the bbox slice will be populated with
+// two points, a minimum then a maximum. Otherwise, the slice is empty.
+//
+// If there is an ID list header, the ids slice will be populated with the
+// IDs from that header. Otherwise, the slice is empty.
+func UnmarshalTWKBWithHeaders(twkb []byte, opts ...ConstructorOption) (g Geometry, bbox []Point, ids []int64, err error) {
+	p := newtwkbParser(twkb, opts...)
+	g, err = p.nextGeometry()
+	if err != nil {
+		return Geometry{}, nil, nil, p.annotateError(err)
+	}
+	if p.hasBBox {
+		bbox, err = UnmarshalTWKBBoundingBoxHeader(twkb)
+		if err != nil {
+			return Geometry{}, nil, nil, p.annotateError(err)
+		}
+	}
+	return g, bbox, p.idList, p.annotateError(err)
+}
+
+// UnmarshalTWKBBoundingBoxHeader checks if the bounding box header
+// exists in the Tiny Well Known Binary (TWKB) and if it exists
+// returns its minimum and maximum points in the bbox slice
+// (otherwise the slice is empty).
+//
+// Because the results are returned as Points, the X, Y, Z, and M values
+// can all be returned. Check the point type to see if the Z and M values
+// are valid.
+//
+// The function returns immediately after parsing the headers.
+// Any remaining geometry is not parsed by this function.
+func UnmarshalTWKBBoundingBoxHeader(twkb []byte) (bbox []Point, err error) {
+	p := newtwkbParser(twkb)
+	bbox, err = p.parseBBoxHeader(twkb)
+	return bbox, p.annotateError(err)
+}
+
+// UnmarshalTWKBEnvelope checks if the bounding box header exists
+// in the Tiny Well Known Binary (TWKB) and iff it exists
+// returns it as an Envelope with a true result.
+//
+// Note that due to the definition of Envelope, only the X and Y
+// coordinates will be returned this way, whereas any Z and M
+// coordinates will be silently ignored by this function.
+//
+// The function returns immediately after parsing the headers.
+// Any remaining geometry is not parsed by this function.
+func UnmarshalTWKBEnvelope(twkb []byte) (bbox Envelope, hasBBox bool, err error) {
+	p := newtwkbParser(twkb)
+	if err = p.parseHeaders(); err != nil {
+		return Envelope{}, false, p.annotateError(err)
+	}
+	if !p.hasBBox {
+		return Envelope{}, false, nil
+	}
+	return newUncheckedEnvelope(
+		XY{
+			p.scalings[0] * float64(p.bbox[0]),
+			p.scalings[1] * float64(p.bbox[2]),
+		},
+		XY{
+			p.scalings[0] * float64(p.bbox[0]+p.bbox[1]),
+			p.scalings[1] * float64(p.bbox[2]+p.bbox[3]),
+		},
+	), true, nil
+}
```

Comment:
The `Envelope` can represent an empty Envelope, so the bool return isn't really needed. Should probably also use the regular envelope constructor rather than the unchecked variant (I don't see any reason not to).

```suggestion
func UnmarshalTWKBEnvelope(twkb []byte) (Envelope, error) {
	p := newtwkbParser(twkb)
	if err := p.parseHeaders(); err != nil {
		return Envelope{}, p.annotateError(err)
	}
	if !p.hasBBox {
		return Envelope{}, nil
	}
	return NewEnvelope([]XY{
		XY{
			p.scalings[0] * float64(p.bbox[0]),
			p.scalings[1] * float64(p.bbox[2]),
		},
		XY{
			p.scalings[0] * float64(p.bbox[0]+p.bbox[1]),
			p.scalings[1] * float64(p.bbox[2]+p.bbox[3]),
		},
	})
}
```

Reply:
I've modified the envelope function as you suggest, fixing the tests and also adjusting the comment to clarify.